### PR TITLE
Sparsity

### PR DESCRIPTION
**Pull request by cnegre**
get sparsity function

### CODE REVIEW

File: src/Fortran-interface/bml_introspection_m.F90 [hunk 1]
```diff
@@ -134,4 +135,22 @@ function bml_get_distribution_mode(a)
...
+  !! @param a The matrix.
+  !! @params threshold The threshold parameter.
+  !! @returns The sparsity of the matrix.
+  function bml_get_sparsity(a, threshold) result(sparsity)
+
+    type(bml_matrix_t), intent(in) :: a
+    real(C_DOUBLE), intent(in) :: threshold
+    real(C_DOUBLE) :: sparsity
+
+    write(111,*)"AAaa", threshold
```
Comment (nicolasbock) [inline src/Fortran-interface/bml_introspection_m.F90]: Debug statement?

File: src/C-interface/bml_sparsity.c [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    const bml_matrix_t * A)
+{
+    const void *A;
+
+    for (int i = 0; i < A->N)
+    {
+        bandwi = bml_get_row_bandwidth(A, i);
+    nnzs = nnzs + bandwi}
+    return 0;
+}
```
Comment (nicolasbock) [inline src/C-interface/bml_sparsity.c]: Why do you need this file? You later add the same function to the introspection routines and this file is never used anywhere.

File: src/C-interface/ellpack/bml_introspection_ellpack_typed.c [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    int A_N = A->N;
+    int A_M = A->M;
+    int *A_nnz = A->nnz;
+
+    for (int i = 0; i < A->N; i++)
+    {
+        // printf("Annz,%d\n", A_nnz[i]);
+        for (int j = 0; j < A_nnz[i]; j++)
+        {
+            if (ABS(A_value[ROWMAJOR(i, j, A_nnz[i], A_N)]) > threshold)
```
Comment (nicolasbock) [inline src/C-interface/ellpack/bml_introspection_ellpack_typed.c]: The `A_value` matrix is `N x M` so this should be `ABS(A_value[ROWMAJOR(i, j, A_N, A_M)])`

File: src/C-interface/ellsort/bml_introspection_ellsort_typed.c [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    int A_M = A->M;
+    int *A_nnz = A->nnz;
+
+    for (int i = 0; i < A->N; i++)
+    {
+        // printf("Annz,%d\n", A_nnz[i]);
+        for (int j = 0; j < A_nnz[i]; j++)
+        {
+            // printf("Avalue,%f\n", A_value[ROWMAJOR(i, j,A_nnz[i], A_N)]);
+            if (ABS(A_value[ROWMAJOR(i, j, A_nnz[i], A_N)]) > threshold)
```
Comment (nicolasbock) [inline src/C-interface/ellsort/bml_introspection_ellsort_typed.c]: Same here: `ABS(A_value[ROWMAJOR(i, j, A_N, A_M)])`

File: src/C-interface/ellsort/bml_introspection_ellsort_typed.c [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            {
+                nnzs++;
+            }
+        }
+    }
+
+    sparsity = (1.0 - (double) nnzs / ((double) (A_N * A_N)));
+
+    return sparsity;
+
```
Comment (nicolasbock) [inline src/C-interface/ellsort/bml_introspection_ellsort_typed.c]: Remove empty line.

File: src/C-interface/ellpack/bml_introspection_ellpack_typed.c [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            {
+                // printf("Annz,%f\n", A_nnz[i]);
+                nnzs++;
+            }
+        }
+    }
+
+    sparsity = (1.0 - (double) nnzs / ((double) (A_N * A_N)));
+    // printf("sparsity,%f\n", sparsity);
```
Comment (nicolasbock) [inline src/C-interface/ellpack/bml_introspection_ellpack_typed.c]: Remove comment.

File: src/C-interface/ellpack/bml_introspection_ellpack_typed.c [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    return sparsity;
+
```
Comment (nicolasbock) [inline src/C-interface/ellpack/bml_introspection_ellpack_typed.c]: Remove empty line.

File: src/C-interface/ellsort/bml_introspection_ellsort_typed.c [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    int j;
+    double sparsity;
+    REAL_T *A_value = (REAL_T *) A->value;
+    int A_N = A->N;
+    int A_M = A->M;
+    int *A_nnz = A->nnz;
+
+    for (int i = 0; i < A->N; i++)
+    {
+        // printf("Annz,%d\n", A_nnz[i]);
```
Comment (nicolasbock) [inline src/C-interface/ellsort/bml_introspection_ellsort_typed.c]: Remove comment.

File: src/C-interface/ellsort/bml_introspection_ellsort_typed.c [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        for (int j = 0; j < A_nnz[i]; j++)
+        {
+            // printf("Avalue,%f\n", A_value[ROWMAJOR(i, j,A_nnz[i], A_N)]);
```
Comment (nicolasbock) [inline src/C-interface/ellsort/bml_introspection_ellsort_typed.c]: Remove comment.

File: tests/get_sparsity_typed.c [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+    if (ABS(sparsity - sparsity_ref) > 1e-12)
+    {
+        LOG_ERROR("bml_get_sparsity is corrupted\n");
+        return -1;
+    }
+
+    printf("Sparsity = %f\n", sparsity);
+
+    bml_deallocate(&A);
```
Comment (nicolasbock) [inline tests/get_sparsity_typed.c]: You have to also deallocate `A_dense`.

File: src/C-interface/dense/bml_introspection_dense_typed.c [hunk 11]
```diff
@@ -65,3 +65,38 @@ int TYPED_FUNC(
...
+ * \param A The bml matrix.
+ * \param threshold The threshold used to compute the sparsity.
+ * \return The sparsity of a matrix.
+ */
+double TYPED_FUNC(
+    bml_get_sparsity_dense) (
+    const bml_matrix_dense_t * A,
+    const double threshold)
+{
+
```
Comment (nicolasbock) [inline src/C-interface/dense/bml_introspection_dense_typed.c]: Please remove empty line

File: src/C-interface/bml_sparsity.h [hunk 12]
```diff
@@ -0,0 +1,14 @@
+/** \file */
```
Comment (nicolasbock) [inline src/C-interface/bml_sparsity.h]: This header file is not used anywhere.

**Comment by nicolasbock** [comment]
@cnegre ping

**Comment by cnegre** [comment]
What should we do to merge this?

File: tests/get_sparsity_typed.c (hunk 10 above)
Comment (cnegre) [inline tests/get_sparsity_typed.c]: done in 7c895c5765e3d29fdd580bce52d8a7db5e3dcc4c commit

File: src/C-interface/dense/bml_introspection_dense_typed.c (hunk 11 above)
Comment (cnegre) [inline src/C-interface/dense/bml_introspection_dense_typed.c]: done in 7c895c5 commit

**Comment by nicolasbock** [comment]
@cnegre I squashed the commits into one. I'll merge the PR once the Travis-CI tests are done.

**Comment by nicolasbock** [comment]
Thanks!

**Pull request merged**